src/common/Utilities/EventMap.h

```diff
@@ -0,0 +1,321 @@
+/*
+ * This file is part of the AzerothCore Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Affero General Public License as published by the
+ * Free Software Foundation; either version 3 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _EVENT_MAP_H_
+#define _EVENT_MAP_H_
+
+#include "Define.h"
+#include "Duration.h"
+#include <map>
+
+class EventMap
+{
+    /**
+    * Internal storage type.
+    * Key: Time as TimePoint when the event should occur.
+    * Value: The event data as uint32.
+    *
+    * Structure of event data:
+    * - Bit  0 - 15: Event Id.
+    * - Bit 16 - 23: Group
+    * - Bit 24 - 31: Phase
+    * - Pattern: 0xPPGGEEEE
+    */
+    typedef std::multimap<uint32, uint32> EventStore;
+
+public:
+    EventMap() { }
+
+    /**
+    * @name Reset
+    * @brief Removes all scheduled events and resets time and phase.
+    */
+    void Reset();
+
+    /**
+     * @name Update
+     * @brief Updates the timer of the event map.
+     * @param time Value to be added to time.
+     */
+    void Update(uint32 time)
+    {
+        _time += time;
+    }
+
+    /**
+    * @name Update
+    * @brief Updates the timer of the event map.
+    * @param time Value in ms to be added to time.
+    */
+    void Update(Milliseconds time)
+    {
+        _time += time.count();
+    }
```

Comment:
What does this output? I have read somewhere that chrono to uint32 can be weird.
Just want to make sure we dont break something

Reply:
You about `.count()` ?

Reply:
yea

Reply:
for `Milliseconds` return `uint64`

Reply:
great

Reply:
i added `static_cast`